core/drivers/imx_ocotp.c

```diff
@@ -0,0 +1,286 @@
+// SPDX-License-Identifier: BSD-2-Clause
+/*
+ * Copyright 2021 NXP
+ */
+#include <arm.h>
+#include <initcall.h>
+#include <mm/core_memprot.h>
+#include <mm/core_mmu.h>
+#include <imx.h>
+#include <io.h>
+#include <drivers/imx_ocotp.h>
+#include <kernel/tee_common_otp.h>
+
+#define OCOTP_CTRL		    0x0
+#define OCOTP_CTRL_ERROR	    BIT32(9)
+#define OCOTP_CTRL_BUSY		    BIT32(8)
+#define OCOTP_SHADOW_OFFSET(_b, _w) ((_b) * (0x40) + (_w) * (0x10) + 0x400)
+
+struct ocotp_instance {
+	unsigned char nb_banks;
+	unsigned char nb_words;
+	TEE_Result (*get_die_id)(uint64_t *ret_uid);
+};
+
+static vaddr_t g_base_addr;
+static struct mutex fuse_read = MUTEX_INITIALIZER;
+static const struct ocotp_instance *g_ocotp;
+
+static TEE_Result ocotp_get_die_id_mx(uint64_t *ret_uid);
+static TEE_Result ocotp_get_die_id_mx7ulp(uint64_t *ret_uid);
```

Comment:
Missed that... normally we try to avoid forward declarations, could be done by moving the struct definitions a bit later in the file. No big deal though.

Reply:
I can move it, no problem

Reply:
OK